### PR TITLE
Addresses #3, added new tests for Search Settings and Related Searches

### PR DESCRIPTION
**Pull request by Marauderer97**
Addresses #3 

Changes: added new tests for Search Settings and Related Searches
NOTE: There can be more tests added for related searches, since there are many input boxes and forms, but I added the minimum, because if we decide to implement the forms differently, or change its classes, we may face problems

Demo Link: Not applicable

Screenshots for the change: 
Not applicable

@harshit98 @nikhilrayaprolu Kindly review

### CODE REVIEW

File: src/app/related-search/related-search.component.spec.ts [hunk 1]
```diff
@@ -75,7 +75,18 @@ describe('RelatedSearchComponent', () => {
...
+  it('should create an instance', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should display results if any, along with right keyword', () => {
+    let compiled = fixture.debugElement.nativeElement;
+    if (component.results.length > 0) {
+      expect(compiled.querySelector('div.card'));
+      expect(component.keyword).not.toBeNull('Keyword has not been initialized');
+    }
```
Comment (harshit98) [inline src/app/related-search/related-search.component.spec.ts]: @Marauderer97 Just asking, did you face the problem `Expect null to be truthy` while testing?

File: src/app/searchsettings/searchsettings.component.spec.ts [hunk 2]
```diff
@@ -1,14 +1,68 @@
...
+import { AdvancedsearchComponent } from '../advancedsearch/advancedsearch.component';
+import { SearchBarComponent } from '../search-bar/search-bar.component';
+import { reducer } from '../reducers/index';
+import { FooterNavbarComponent } from '../footer-navbar/footer-navbar.component';
+import { AboutComponent } from '../about/about.component';
+import { ModalComponent } from 'ng2-bs3-modal/ng2-bs3-modal';
+import { ContactComponent } from '../contact/contact.component';
+import {InfoboxComponent} from "../infobox/infobox.component";
+import { RelatedSearchComponent } from '../related-search/related-search.component';
+import {KnowledgeapiService} from "../knowledgeapi.service";
```
Comment (harshit98) [inline src/app/searchsettings/searchsettings.component.spec.ts]: There was no need to import all the modules. You are importing unnecessary modules.

**Comment by harshit98** [comment]
@Marauderer97 Yes, the test which you wrote looks sufficient. Kindly go through requested changes. :)

File: src/app/related-search/related-search.component.spec.ts (hunk 1 above)
Comment (Marauderer97) [inline src/app/related-search/related-search.component.spec.ts]: @harshit98, no faced no such problem.

File: src/app/searchsettings/searchsettings.component.spec.ts (hunk 2 above)
Comment (Marauderer97) [inline src/app/searchsettings/searchsettings.component.spec.ts]: @harshit98 I imported all modules because otherwise, my tests kept being failed, and besides, in results and index, all these modules were imported. I also thought they will be useful for future use, but if you insist I can remove them.

**Comment by Marauderer97** [comment]
@nikhilrayaprolu, the Travis build was not passing, so I imported all possible modules, but ng test still fails, showing errors similar to this:
image
What went wrong with the app-intelligence component?

**Comment by nikhilrayaprolu** [comment]
sure I will work on this now, and push the changes to your branch.

File: src/app/searchsettings/searchsettings.component.spec.ts [hunk 3]
```diff
@@ -1,14 +1,68 @@
...
+        FooterNavbarComponent,
+        AboutComponent,
+        ContactComponent,
+        ModalComponent,
+        InfoboxComponent,
+        RelatedSearchComponent,
+        AutoCompleteComponent,
+        ThemeComponent,
+        DropdownComponent,
+        SearchsettingsComponent
```
Comment (nikhilrayaprolu) [inline src/app/searchsettings/searchsettings.component.spec.ts]: image

add IntelligenceComponent here this will solve the issue.

File: src/app/related-search/related-search.component.spec.ts [hunk 4]
```diff
@@ -77,7 +77,18 @@ describe('RelatedSearchComponent', () => {
...
   });
 
-  it('should create', () => {
+  it('should create an instance', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should display results if any, along with right keyword', () => {
+    let compiled = fixture.debugElement.nativeElement;
+    if (component.results.length > 0) {
```
Comment (nikhilrayaprolu) [inline src/app/related-search/related-search.component.spec.ts]: where are results initialised onto a component instance.
if no results are initialised these tests won't run at all, AFAIK

File: src/app/searchsettings/searchsettings.component.spec.ts (hunk 2 above)
Comment (nikhilrayaprolu) [inline src/app/searchsettings/searchsettings.component.spec.ts]: @harshit98 do you have any suggestions regarding this? any alternative to remove all imports.

File: src/app/searchsettings/searchsettings.component.spec.ts [hunk 5]
```diff
@@ -19,4 +75,41 @@ describe('SearchsettingsComponent', () => {
     fixture.detectChanges();
   });
 
+
+  it('should create an instance', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should create a FooterNavbar Component', () => {
+    const footerNavbar = new FooterNavbarComponent();
```
Comment (nikhilrayaprolu) [inline src/app/searchsettings/searchsettings.component.spec.ts]: what is the significance of this test?

**Comment by harshit98** [comment]
@Marauderer97 In searchsetting component, import the following: 
- FooterNavbar
- RouterTestingModule
- reducer
- StoreModule (and don't forget to call them in imports as StoreModule.provideStore(reducer))

Then then remove the rest modules. If `ng test` says something is missing, add that module.

The most common errors `ng test` can throw like:
- no provider for http (import HttpModule in this case)
- they can also be related to service. If following service is not being provided, add that too. (in this case, it won't be asking though)

**Comment by harshit98** [comment]
@Marauderer97 Import these modules only, it will pass the test - https://github.com/harshit98/susper.com/commit/5b70169c05d0f18bd0269b67bb364fc47f2e2f2a

File: src/app/related-search/related-search.component.spec.ts (hunk 4 above)
Comment (Marauderer97) [inline src/app/related-search/related-search.component.spec.ts]: I didnot understand, where do you wish I should initialize results?

Comment (Marauderer97) [inline src/app/related-search/related-search.component.spec.ts]: In the jasmine official tutorial, there was no mention of initializing, he said you can access values of the component using its name, that is why I did this.

File: src/app/searchsettings/searchsettings.component.spec.ts (hunk 5 above)
Comment (Marauderer97) [inline src/app/searchsettings/searchsettings.component.spec.ts]: It ensures the presence of a footernavbar component on the page

**Comment by Marauderer97** [comment]
@harshit98 Alright, I am waiting for @nikhilrayaprolu to approve the other tests, will then add a commit with all requested changes including removing extra modules

File: src/app/related-search/related-search.component.spec.ts (hunk 4 above)
Comment (nikhilrayaprolu) [inline src/app/related-search/related-search.component.spec.ts]: image

this will give default results without initialising, i.e we have results=[] initialized to empty array. due to which the tests won't run at all.
actually, we should avoid conditional statements while writing tests.

Comment (nikhilrayaprolu) [inline src/app/related-search/related-search.component.spec.ts]: just try removing the conditional statement, you will find that tests fail. everytime. 
solution to this is:-
1) directly initialize resutls variable as shown in above image.
2) have a spy on search-service ( https://angular.io/guide/testing#test-a-component-with-an-async-service ) and get results by spying upon that

File: src/app/searchsettings/searchsettings.component.spec.ts (hunk 5 above)
Comment (nikhilrayaprolu) [inline src/app/searchsettings/searchsettings.component.spec.ts]: this is no way a test for instance if I import a results component and write
```
const resultscomponent = new ResultsComponent();
expect(resultscomponent).toBeTruthy();

```
this test gets passed.
even though searchsettings component doesn't have a results component in it.

**Comment by harshit98** [comment]
@Marauderer97 any status regarding this PR ?

**Comment by Marauderer97** [comment]
@harshit98 In the tutorial, Jasmine requires a function to spy on, however, here we get results from fromRoot.getKnowledge and that is giving an error in the Jasmine tests, I have incorporated the rest of the changes

**Comment by Marauderer97** [comment]
@harshit98 I imported only the modules you said, travis is failing

**Comment by Marauderer97** [comment]
@nikhilrayaprolu All checks have passed, please approve now :sweat_smile:

**Comment by Marauderer97** [comment]
@nikhilrayaprolu I have tried to initialize the spy service as you suggested, but the method call raises errors, please consider this an iteration, and add to it

File: src/app/related-search/related-search.component.spec.ts [hunk 6]
```diff
@@ -79,7 +50,19 @@ describe('RelatedSearchComponent', () => {
...
 
-  it('should create', () => {
+  it('should create an instance', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should display results if any, along with right keyword', () => {
+    let knowledgeService = fixture.debugElement.injector.get(KnowledgeapiService);
+    let compiled = fixture.debugElement.nativeElement;
+    if (component.results.length > 0) {
```
Comment (nikhilrayaprolu) [inline src/app/related-search/related-search.component.spec.ts]: any update regarding this?
A test should have a definite path, we should avoid having condtionals inside tests.
I remembered commenting regarding this at https://github.com/fossasia/susper.com/pull/575#pullrequestreview-47794656 and gave suggestions as below:
just try removing the conditional statement, you will find that tests fail, everytime. 
solution to this is:-
1) directly initialize results variable as shown in above image.
2) have a spy on search service ( https://angular.io/guide/testing#test-a-component-with-an-async-service ) and get results by spying upon that

Comment (Marauderer97) [inline src/app/related-search/related-search.component.spec.ts]: @nikhilrayaprolu Like I said, the spy service gives errors since it is being called on the constructor function, I donot know how to prevent it, in the tutorial you have mentioned, the spy function requires a method to be called on
